:: is not a class type

```scala
import scala.collection.immutable.::
object A {
  def main(args: Array[String]): Unit = {
     new ::(args, Nil)
  }
}
```

```
A.scala:4: error: :: is not a class type
     new ::(args, Nil)
         ^
one error found
```

--- a/src/dotty/tools/dotc/typer/Typer.scala
+++ b/src/dotty/tools/dotc/typer/Typer.scala
@@ -347,7 +347,12 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
         val clsDef = TypeDef(x, templ).withFlags(Final)
         typed(cpy.Block(tree)(clsDef :: Nil, New(Ident(x), Nil)), pt)
       case _ =>
-        val tpt1 = typedType(tree.tpt)
+        var tpt1 = typedType(tree.tpt)
+        if (tpt1.tpe.isHK) {
+          val deAliased = tpt1.tpe.dealias.EtaReduce
+          if (deAliased.exists && deAliased.ne(tpt1.tpe))
+            tpt1 = tpt1.withType(deAliased)
+        }
         checkClassTypeWithStablePrefix(tpt1.tpe, tpt1.pos, traitReq = false)
         assignType(cpy.New(tree)(tpt1), tpt1)
         // todo in a later phase: checkInstantiatable(cls, tpt1.pos)

--- a/tests/pos/i815.scala
+++ b/tests/pos/i815.scala
@@ -1,0 +1,11 @@
+import scala.collection.immutable.::
+class C[T](x: T)
+object A {
+  def main(args: Array[String]): Unit = {
+    val x = new C("A")
+    val y = new ::(args, Nil)
+    val z = ::(args, Nil)
+    println(y)
+    println(z)
+  }
+}

